src/libraries/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs

```diff
@@ -450,6 +451,139 @@ public static void CustomTrustModeWithNoCustomTrustCerts()
             }
         }
 
+        [Fact]
+        public static void NameConstraintViolation_PermittedTree_Dns()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+            builder.AddDnsName("microsoft.com");
+
+            // permitted DNS name constraint for .example.com
+            string nameConstraints = "3012A010300E820C2E6578616D706C652E636F6D";
+
+            TestNameConstrainedChain(nameConstraints, builder, (bool result, X509Chain chain) => {
+                Assert.False(result, "chain.Build");
+                Assert.Equal(PlatformNameConstraints(X509ChainStatusFlags.HasNotPermittedNameConstraint), chain.AllStatusFlags());
+            });
+        }
+
+        [Fact]
+        public static void NameConstraintViolation_ExcludedTree_Dns()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+            builder.AddDnsName("www.example.com");
+
+            // excluded DNS name constraint for example.com.
+            string nameConstraints = "3012A110300E820C2E6578616D706C652E636F6D";
+
+            TestNameConstrainedChain(nameConstraints, builder, (bool result, X509Chain chain) => {
+                Assert.False(result, "chain.Build");
+                Assert.Equal(PlatformNameConstraints(X509ChainStatusFlags.HasExcludedNameConstraint), chain.AllStatusFlags());
+            });
+        }
+
+        [Fact]
+        [PlatformSpecific(~TestPlatforms.OSX)] // macOS appears to just completely ignore min/max.
+        public static void NameConstraintViolation_PermittedTree_HasMin()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+            builder.AddDnsName("example.com");
+
+            // permitted DNS name constraint for example.com with a MIN of 9.
+            string nameConstraints = "3015A0133011820C2E6578616D706C652E636F6D800109";
+
+            TestNameConstrainedChain(nameConstraints, builder, (bool result, X509Chain chain) => {
+                Assert.False(result, "chain.Build");
+                Assert.Equal(PlatformNameConstraints(X509ChainStatusFlags.HasNotSupportedNameConstraint), chain.AllStatusFlags());
+            });
+        }
+
+        [Fact]
+        [PlatformSpecific(~TestPlatforms.Windows)] // Windows seems to skip over nonsense GeneralNames.
```

Comment:
We can't easily make Linux match Window's behavior. Consider a certificate that had name constraint A and B. B's NC is nonsense and no reasonable chain builder could do anything intelligent with it.

CAPI will enforce A, and skip B.

OpenSSL will report `X509_V_ERR_UNSUPPORTED_NAME_SYNTAX` for the whole operation. If we treat this error as `NoError`, then we are reported `NoError` for both A and B constraints, whereas Windows may have at least checked A.

To err on the side of not letting something through that we shouldn't, we report the error.